appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go

```diff
@@ -0,0 +1,577 @@
+	log := log.FromContext(ctx).WithValues("request", req)
+
+	dtc := applicationv1alpha1.DeploymentTargetClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(&dtc), &dtc); err != nil {
+		// Don't requeue if the requested object is not found/deleted.
+		if apierr.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Add the deletion finalizer if it is absent.
+	if addFinalizer(&dtc, sharedutil.FinalizerBinder) {
+		if err := r.Client.Update(ctx, &dtc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer %s to DeploymentTargetClaim %s in namespace %s: %v", sharedutil.FinalizerBinder, dtc.Name, dtc.Namespace, err)
+		}
+	}
+
+	// Handle deletion if the DTC has a deletion timestamp set.
+	if dtc.GetDeletionTimestamp() != nil {
+		// If the DTC is bound set the status of the corresponding DT to Released
+		if isDTCBound(dtc) {
+			dt, err := getDTBoundByDTC(ctx, r.Client, &dtc)
+			if err != nil && !apierr.IsNotFound(err) {
+				return ctrl.Result{}, err
+			}
+
+			if dt != nil {
+				err = updateDTStatusPhase(ctx, r.Client, dt, applicationv1alpha1.DeploymentTargetPhase_Released)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		}
+
+		if removeFinalizer(&dtc, sharedutil.FinalizerBinder) {
+			if err := r.Client.Update(ctx, &dtc); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer %s from DeploymentTargetClaim %s in namespace %s: %v", sharedutil.FinalizerBinder, dtc.Name, dtc.Namespace, err)
+			}
+		}
+
+		log.Info("Removed finalizer from DeploymentTargetClaim")
+		return ctrl.Result{}, nil
+	}
+
+	// If the binding is alredy done, we need to check if the DTC is still bound to a DT
+	// and update the status accordingly
+	if isBindingCompleted(dtc) {
+		if err := handleBoundedDeploymentTargetClaim(ctx, r.Client, dtc, log); err != nil {
+			log.Error(err, "failed to process bounded DeploymentTargetClaim")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// If the user doesn't set the DT, check if there is a matching DT available
+	// or if it needs to be dynamically provisioned.
+	if dtc.Spec.TargetName == "" {
+		dt, err := findMatchingDTForDTC(ctx, r.Client, dtc)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// If a best match DT is available bind it to the current DTC.
+		if dt != nil {
+			err = bindDeploymentTargetCliamToTarget(ctx, r.Client, &dtc, dt, true)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			if err := updateDTStatusPhase(ctx, r.Client, dt, applicationv1alpha1.DeploymentTargetPhase_Bound); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			if err := updateDTCStatusPhase(ctx, r.Client, &dtc, applicationv1alpha1.DeploymentTargetClaimPhase_Bound); err != nil {
+				return ctrl.Result{}, err
+			}
```

Comment:
This scenario is handled by the [handleBoundedDeploymentTargetClaim()](https://github.com/redhat-appstudio/managed-gitops/blob/972b6dd3b095cee7f9dbd39f028b6aef598c8907/appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go#L116) function in the reconciler. If a DTC has a `bind-complete` annotation the reconciler checks:
1. If the corresponding DT is present. If not, it sets the DTC status to lost.
2. Updates the status of both DT and DTC back to `Bound`.

So if a DTC is already bounded and the status update fails or deviates due to some reason, it will be synced in the next reconciliation cycle.

There is a [unit test](https://github.com/redhat-appstudio/managed-gitops/blob/972b6dd3b095cee7f9dbd39f028b6aef598c8907/appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller_test.go#L214) for covering this case.